url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/example/KalmanFilterAlg.java

```diff
@@ -0,0 +1,141 @@
+import org.ejml.factory.LinearSolverFactory;
+import org.ejml.interfaces.linsol.LinearSolver;
+
+import static org.ejml.ops.CommonOps.*;
+
+/**
+ * The difference between this and {@link KalmanFilterOps} is that it takes advantage of
+ * the covariance matrix being a symetric positive semi-definite matrix.  This allows
+ * it to be decomposed using {@link org.ejml.alg.dense.decomposition.chol.CholeskyDecompositionInner_D64}.  There are two advantages here,
+ * 1) all memory is predeclared and 2) CholeksyDecomposition is more efficient than the more
+ * generic {@link org.ejml.alg.dense.decomposition.lu.LUDecompositionAlt_D64 LUDecomposition}.  It also makes
+ * calls to matrix vector multiplcation operations, which has a slight performance advantage.
+ *
+ * @author Peter Abeles
+ */
+public class KalmanFilterAlg implements KalmanFilter{
+
+    // kinematics description
+    private DenseMatrix64F F;
+    private DenseMatrix64F Q;
+    private DenseMatrix64F H;
+
+    // sytem state estimate
+    private DenseMatrix64F x;
+    private DenseMatrix64F P;
+
+    // these are predeclared for efficency reasons
+    private DenseMatrix64F a,b;
+    private DenseMatrix64F y,S,S_inv,c,d;
+    private DenseMatrix64F K;
+
+    private LinearSolver<DenseMatrix64F> solver;
+
+    @Override
+    public void configure(DenseMatrix64F F, DenseMatrix64F Q, DenseMatrix64F H) {
+        this.F = F;
+        this.Q = Q;
+        this.H = H;
+
+        int dimenX = F.numCols;
+        int dimenZ = H.numRows;
+
+        a = new DenseMatrix64F(dimenX,1);
+        b = new DenseMatrix64F(dimenX,dimenX);
+        y = new DenseMatrix64F(dimenZ,1);
+        S = new DenseMatrix64F(dimenZ,dimenZ);
+        S_inv = new DenseMatrix64F(dimenZ,dimenZ);
+        c = new DenseMatrix64F(dimenZ,dimenX);
+        d = new DenseMatrix64F(dimenX,dimenZ);
+        K = new DenseMatrix64F(dimenX,dimenZ);
+
+        // covariance matrices are symmetric positive semi-definite
+        solver = LinearSolverFactory.symmPosDef(dimenX);
+        // wrap the solver so that it doesn't modify the input
+//        solver = new LinearSolverSafe<DenseMatrix64F>(solver);
+        // A little bit more performance can be gained by letting S be modified.  In some
+        // applications S should not be modified.
+
+        x = new DenseMatrix64F(dimenX,1);
+        P = new DenseMatrix64F(dimenX,dimenX);
+    }
+
+    @Override
+    public void setState(DenseMatrix64F x, DenseMatrix64F P) {
+        this.x.set(x);
+        this.P.set(P);
+    }
+
+    @Override
+    public void predict() {
+
+        // x = F x
+        MatrixVectorMult.mult(F,x,a);
+        x.set(a);
+
+        // P = F P F' + Q
+        MatrixMatrixMult.mult_small(F,P,b);
+        MatrixMatrixMult.multTransB(b,F, P);
+        addEquals(P,Q);
+    }
+
+    @Override
+    public void update(DenseMatrix64F z, DenseMatrix64F R) {
+        // y = z - H x
+        MatrixVectorMult.mult(H,x,y);
+        sub(z,y,y);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;z&#39; (index: 0 | type: DenseMatrix64F). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;R&#39;. The same parameter type is validated here for example: url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/decomposition/bidiagonal/BenchmarkBidiagonalDecomposition.java:41. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).